src/network_engine.cpp

```diff
@@ -0,0 +1,875 @@
+    }
+
+    Blob bnodes6;
+    if (want & WANT6) {
+        bnodes6.resize(NODE6_INFO_BUF_LEN * TARGET_NODES);
+        bnodes6.resize(NODE6_INFO_BUF_LEN * buff(bnodes6, id, nodes6));
+    }
+
+    return {std::move(bnodes4), std::move(bnodes6)};
+}
+
+std::shared_ptr<NetworkEngine::Request>
+NetworkEngine::sendListen(std::shared_ptr<Node> n, const InfoHash& infohash, const Blob& token,
+        RequestCb on_done, RequestExpiredCb on_expired) {
+    auto tid = TransId {TransPrefix::LISTEN, getNewTid()};
+    msgpack::sbuffer buffer;
+    msgpack::packer<msgpack::sbuffer> pk(&buffer);
+    pk.pack_map(5);
+
+    pk.pack(std::string("a")); pk.pack_map(3);
+      pk.pack(std::string("id"));    pk.pack(myid);
+      pk.pack(std::string("h"));     pk.pack(infohash);
+      pk.pack(std::string("token")); packToken(pk, token);
+
+    pk.pack(std::string("q")); pk.pack(std::string("listen"));
+    pk.pack(std::string("t")); pk.pack_bin(tid.size());
+                               pk.pack_bin_body((const char*)tid.data(), tid.size());
+    pk.pack(std::string("y")); pk.pack(std::string("q"));
+    pk.pack(std::string("v")); pk.pack(my_v);
+
+
+    Blob b {buffer.data(), buffer.data() + buffer.size()};
+    std::shared_ptr<Request> req(new Request {tid.getTid(), n, std::move(b),
+        [=](std::shared_ptr<Request> req_status, ParsedMessage&& msg) { /* on done */
+            if (on_done)
+                on_done(req_status, deserializeNodesValues(msg));
+        },
+        [=](std::shared_ptr<Request> req_status, bool) { /* on expired */
+            if (on_expired)
+                on_expired(req_status, {});
+        },
+        true
+    });
+    sendRequest(req);
+    ++out_stats.listen;
+    return req;
+}
+
+void
+NetworkEngine::sendListenConfirmation(const sockaddr* sa, socklen_t salen, TransId tid) {
+    msgpack::sbuffer buffer;
+    msgpack::packer<msgpack::sbuffer> pk(&buffer);
+    pk.pack_map(4);
+
+    pk.pack(std::string("r")); pk.pack_map(2);
+      pk.pack(std::string("id")); pk.pack(myid);
+      insertAddr(pk, sa, salen);
+
+    pk.pack(std::string("t")); pk.pack_bin(tid.size());
+                               pk.pack_bin_body((const char*)tid.data(), tid.size());
+    pk.pack(std::string("y")); pk.pack(std::string("r"));
+    pk.pack(std::string("v")); pk.pack(my_v);
+
+    send(buffer.data(), buffer.size(), 0, sa, salen);
+}
+
+std::shared_ptr<NetworkEngine::Request>
+NetworkEngine::sendAnnounceValue(std::shared_ptr<Node> n, const InfoHash& infohash, const Value& value, time_point created,
+        const Blob& token, RequestCb on_done, RequestExpiredCb on_expired) {
+    auto tid = TransId {TransPrefix::ANNOUNCE_VALUES, getNewTid()};
+    msgpack::sbuffer buffer;
+    msgpack::packer<msgpack::sbuffer> pk(&buffer);
+    pk.pack_map(5);
+
+    pk.pack(std::string("a")); pk.pack_map((created < scheduler.time() ? 5 : 4));
+      pk.pack(std::string("id"));     pk.pack(myid);
```

Comment:
This indentation came from the previous code written by @aberaud and I think this is intentional and it means to show that packing is made under a same key in the map structure. @aberaud, is it right?